### ISSUE TITLE
Add support for point geometries source data to "import culverts/weirs/orifices"

### ISSUE THREAD

**Issue by leendertvanwolfswinkel** [nens/threedi-schematisation-editor#167]
I would like to release the "import point structures" tool that was developed as part of the MIKE importer separately. For users, Integrating it in the "Import culverts" algorithm would make workflows for users the easiest. I propose to add five arguments to the "conversion settings" section of the config JSON (and to the GUI once that is finished):

```
{
  "target_layer": "culvert",
  "conversion_settings": {
    "use_snapping": true,
    "snapping_distance": 0.5,
    "create_connection_nodes": true, 
    "length_source_field": "duiker_lengte", NEW
    "length_fallback_value": 10, NEW
    "azimuth_source_field": "duiker_richting", NEW
    "azimuth_fallback_value": 90, NEW
    "edit_channels": true NEW
  },
```

- The `length_source_field` contains the name of the field in the source layer that contains the length that the line feature should have. This can also be `None`. This setting is ignored when the input has a line geometry.
- The `length_fallback_value` is a float value to be used when `length_source_field` is `None` or that attribute is NULL for this specific feature. This setting is ignored when the input has a line geometry.
- If `edit_channels` is `true` and the source layer has a point geometry, the behaviour is as in the algorithm that you wrote for the mike converter.
- If `edit_channels` is `true` and the source layer has a linestring geometry, and the start and end vertices are both within snapping distance of the same channel, it should work the same as in the algorithm that you wrote for the mike converter. otherwise, use the same logic as already exists in the culvert importer
- If `edit_channels` is `false` and the source layer has a point geometry, use the value in `azimuth_source_field` or if not available `azimuth_fallback_value` to draw a line. After having the line, the same logic can be applied as already exists in the culvert importer
- If `edit_channels` is `false` and the source layer has a line geometry, use the same logic as already exists in the culvert importer

- [ ] Also update the UI

When this has been implemented, we can simply create config files that work for the mike converter intermediate outputs

**Comment by ldebek** [comment]
Testing data would be very helpful to properly test this functionality.

**Comment by leendertvanwolfswinkel** [comment]
@ldebek sorry this took so long... this zip contains:
- Dommelerwaard import HyDamo.gpkg --> 3Di Schematisation
- HyDAMO_Dommelerwaard.gpkg --> source data. Use layer "Brug" to import as culverts, layer "Stuw" to import as weir

[Dommelerwaard.zip](https://github.com/user-attachments/files/16020476/Dommelerwaard.zip)

**Comment by leendertvanwolfswinkel** [comment]
@ldebek I tested it with the test data and this config json to import the layer "Stuw" as weir. Feedback:
- [ ] The "Length fallback value" is now prepopulated with 10 m, please set this to 1 m instead 
- The channels are not cut even though I checked the "edit channels" box
- <img width="1075" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMnpQQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--751c58efb5ba19aea16ef436a784b6d3878d7ac8/image.png" alt="image.png" />
-  If I add an extra weir to the source data that is close to another weir (< length), they are overlapping. I think I remember the algorithm for the MIKE importer took care of this? Perhaps related to the "edit channels" thing not working?
-  If the weir is close to the start or end of the channel, it may result in short segments (see image below). Also for this, iirc this was also taken care of in the algorithm for the MIKE importer? Perhaps this is also dysfunctional because the whole "edit channels" thing does not work? If so, what tolerance is used, i.e. how short is a cut-up channel allowed to be? I would propose to regard any channel segment < 5 m as "too short" and shift the weir to the start/end of the channel in such cases
-   <img width="981" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMy9QQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--89a1d5a2a14f92df3cf05b7304184e624c924a67/image.png" alt="image.png" />

**Comment by leendertvanwolfswinkel** [comment]
Hi @ldebek I did some further testing with the new version. The channel editing function now works. I have some further wishes:
- [ ] Improve the way the importer handles structures at junctions. Currently, it creates a structure on each channel that is intersected. So in this case, I get three weirs for the price of one:
<img width="491" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMlBoQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fcc7780a24b59384ce72a8f45c740f323d177c00/image.png" alt="image.png" />
 
The best would be to handle them as follows:
- Snap the point only to the nearest channel
- If there are > 1 nearest channels (exact same distance), do not process the point
- When the import is finished, but some points have been skipped, show a warning telling the user which points where skipped. Would be extra nice if a point layer of "skipped points" could be added to the project

- [ ] When the channel is split, some parts of the channel may end up without a cross-section location. If cutting up the channel results in channels without cross-section locations, please copy the nearest cross-section location from the original channel to that new channel: <img width="1156" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMjNoQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--96fe70cc74d681314c6797bd6c2a37660695d865/image.png" alt="image.png" />

- [ ] Pipes should be treated as channels, not as structures.
 - Remove the new options from the Pipes importer
 - Add an option "edit pipes" to importers for weir and orifice (not relevant for culverts)
 - If a weir or orifice point is on a pipe (or snapped to the pipe), it can be handled a bit more simply than for channels; just find if it is closer to the start or to the end of the pipe, and insert the weir/orifice there. If the length of the weir/orifice is more than the length of the pipe, the entire pipe can be replaced.
 <img width="891" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM25oQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e5c706b657d3553318bb4f701ea30f9d122d2907/image.png" alt="image.png" />

**Comment by ldebek** [comment]
@leendertvanwolfswinkel Proposed handling of structures at junctions will require significant refactoring of currently used alghorithm and adds another level of complexity for handling all the edge cases possible on T junctions. I propose to prepare a list of those edge cases with desired solutions and move that wishes to the separate ticket. For now I just added a few extra lines to avoid creating multiple structures from a single input point.
P.S. Changes for the channels without cross-section location and removing new options from the Pipe importer are already available on the working branch.

**Comment by leendertvanwolfswinkel** [comment]
@ldebek 

OK, wrt the junctions, I will have a note about this added in the documentation. Let's see how it goes and if we get any user feedback on this issue; if so, we can always add more complexity later.

Wrt to the channels without cross-section location: I tested this and it works nicely. However I did notice this:

<img width="467" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL1I2Qmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fc5f56da1c282a913e12ad64d62b15363238c956/image.png" alt="image.png" />

The part of the channel where the original cross-section location is, was deleted for inserting the weir. So the a new cross-section location is created on the remaining part of the channel. However, the original cross-section location is still there, which is not valid for 3Di. Would it be possible to automatically remove cross-section locations that no longer are located on a channel?

**Comment by ldebek** [comment]
@leendertvanwolfswinkel Added removal of the cross-section locations that no longer are located on a channel.

**Comment by leendertvanwolfswinkel** [comment]
@ldebek I think your removal routine is a bit too trigger happy :D . It solved some issues, but created new ones. Some cross-section locations that should not have been removed were removed. See the image



 <img width="839" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeEdUQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f68fd5d5d889bf1c820502a508ff19d4456c6205/before%20and%20after.png" alt="before and after.png" />

**Comment by ldebek** [comment]
@leendertvanwolfswinkel Please check if problematic cross-section locations have correct channel_id assigned. I had similar issues when channel_id of cross section does not match spatial relation.

**Comment by leendertvanwolfswinkel** [comment]
Hi @ldebek I checked, but before the import all cross-section locations' channel_ids match the id of the channel they are on geometrically. 

Locations of the two cases that are wrong after the most recent fix:

Top case: between "Stuw" with FID 22 and 17
Case on 3rd row: to the west of Stuw 14

**Comment by ldebek** [comment]
@leendertvanwolfswinkel Could you please share with me the data you are using for testing?

**Comment by leendertvanwolfswinkel** [comment]
Same as in one of the comments above (Dommelerwaard)

**Comment by ldebek** [comment]
@leendertvanwolfswinkel Fix pushed.

**Comment by leendertvanwolfswinkel** [comment]
@ldebek just one last thing...

no just kidding please merge this please get it over with

great stuff
